File: backend/app/core/policy.py
from pathlib import Path
import re, yaml
from app.core.config import CONFIG_DIR

class PolicyGate:
    def __init__(self):
        self.policy = yaml.safe_load((Path(CONFIG_DIR) / "agent-policy.yaml").read_text())
        self.blocked_keywords = [x.lower() for x in self.policy.get("blocked_keywords", [])]

    def check_text(self, text: str):
        lowered = text.lower()
        for keyword in self.blocked_keywords:
            if keyword in lowered:
                return {"allowed": False, "reason": f"blocked by safety policy: {keyword}"}
        return {"allowed": True, "reason": "allowed"}

    def check_mode(self, mode: str):
        allowed = set(self.policy.get("allowed_categories", []))
        if mode not in allowed:
            return {"allowed": False, "reason": f"mode {mode} is not allowed"}
        return {"allowed": True, "reason": "allowed"}

File: backend/src/core/policy.ts
import { readFileSync } from "node:fs";
import path from "node:path";
import { env } from "../lib/env.js";
import { parsePolicyYaml } from "../lib/yaml.js";

type PolicyConfig = {
  allowed_categories?: string[];
  blocked_keywords?: string[];
};

export class PolicyGate {
  private readonly config: PolicyConfig;
  private readonly blockedKeywords: string[];

  constructor() {
    const filePath = path.join(env.configDir, "agent-policy.yaml");
    this.config = parsePolicyYaml(readFileSync(filePath, "utf8")) as PolicyConfig;
    this.blockedKeywords = (this.config.blocked_keywords ?? []).map((item) => item.toLowerCase());
  }

  checkText(text: string) {
    const lowered = text.toLowerCase();
    for (const keyword of this.blockedKeywords) {
      if (lowered.includes(keyword)) {
        return { allowed: false, reason: `blocked by safety policy: ${keyword}` };
      }
    }
    return { allowed: true, reason: "allowed" };
  }

  checkMode(mode: string) {
    const allowed = new Set(this.config.allowed_categories ?? []);
    if (!allowed.has(mode)) {
      return { allowed: false, reason: `mode ${mode} is not allowed` };
    }
    return { allowed: true, reason: "allowed" };
  }
}
